### PR TITLE
Handshake test

### PR DESCRIPTION
**Pull request by themighty1**
Does not compile. Just a draft branch to get early feedback and then polish as needed.

### CODE REVIEW

File: tls-client/src/handshaker.rs [hunk 1]
```diff
@@ -1,39 +1,234 @@
...
+        let mut a1_seed = [0u8; 79];
+        a1_seed[..32].copy_from_slice(&a1);
+        a1_seed[32..].copy_from_slice(&seed);
+        let mut verify_data = [0u8; 12];
+        verify_data.copy_from_slice(&hmac_sha256(&ms, &a1_seed)[..12]);
+        let mut client_finished = [0u8; 16];
+        client_finished[..4].copy_from_slice(&[0x14, 0x00, 0x00, 0x0C]);
+        client_finished[4..].copy_from_slice(&verify_data);
+
+        // encrypt client finished
```
Comment (sinui0) [inline tls-client/src/handshaker.rs:154]: This function does not encrypt the ClientFinished, the return value should simply be the verify_data bytes. The tls-client will construct the plaintext CF and pass it to encrypt()

**Review by sinui0 — COMMENTED** [review]
Looks like a good start!

I will also add some additional methods to the trait along the lines of:
`fn select_protocol_version(&mut self, version: ProtocolVersion) -> Result<(), Error>`
and
`fn select_cipher_suite(&mut self, suite: SupportedCipherSuite) -> Result<(), Error>`

which tls-client will call to signal what has been selected in the handshake.

File: tls-client/src/handshaker.rs [hunk 2]
```diff
@@ -1,39 +1,234 @@
...
+
+        Ok(())
     }
-    async fn set_server_key_share(&mut self, _key: PublicKey) -> Result<(), Self::Error> {
-        Err(Error::General("handshaker not yet available".to_string()))
+
+    async fn client_key_share(&mut self) -> Result<PublicKey, Error> {
+        // return our ecdh pubkey
+        Ok(PublicKey {
+            group: NamedGroup::secp256r1,
```
[the remark this answers is not in the document]
Comment (themighty1) [reply tls-client/src/handshaker.rs:120]: No, we may add  other curves in the future.

File: tls-client/src/handshaker.rs [hunk 3]
```diff
@@ -1,39 +1,234 @@
...
+use aes_gcm::aead::{generic_array::GenericArray, Aead, NewAead, Payload};
+use aes_gcm::Aes128Gcm;
+use p256::{ecdh::EphemeralSecret, EncodedPoint, PublicKey as ECDHPublicKey};
+use rand::rngs::OsRng;
+use rand::{thread_rng, Rng};
 use tls_aio::{
     cipher::{MessageDecrypter, MessageEncrypter},
     handshaker::Handshake,
 };
+use tls_core::handshake::utils::{hmac_sha256, seed_ke, seed_ms};
```
[the remark this answers is not in the document]
Comment (themighty1) [reply tls-client/src/handshaker.rs:11]: it can def be used for tls13 but tls12's prf does not map to what hkdf does.

**Pull request closed without merging**